### ISSUE TITLE
Decide semantics of logical operators and incompatible types

### ISSUE THREAD

**Issue by erikgrinaker** [sanity-io/gradient#563]
Consider the following:

```
gradient> true && 1
                   1
gradient> "string" || false
                           "string"
```

If Gradient should be strongly typed (which I prefer), these should error out. Related to #560 and #561 .

**Comment by erikgrinaker** [comment]
This also goes for comparisons between different types, and passing unexpected types to functions, e.g.:

```
gradient> "a" > 1
                 false
gradient> "a" < 1
                 false
```

We should really decide once and for all if Gradient is strongly or weakly typed, and map out the implications if weakly (e.g. truthiness and comparison semantics).

**Comment by erikgrinaker** [comment]
A complication here is that we allow mixed types in arrays and documents , e.g. `[1, "a", true]` and `"value": 1`, `"value": "a"`. We have to allow ordering of these, and since all sorting algorithms fundamentally rely on comparing two arbitrary values in the set, we must specify well-defined comparison operations between all types. This probably precludes strong typing. 😢

**Comment by erikgrinaker** [comment]
Actually, we can't have parity between conparison operators and sorting functions either. We probably want `null < 1` to yield `null`, but still want to order nulls first when sorting. We can possibly get away with sorting by datatype first.

Suggested order: nulls, booleans, numbers, strings, arrays, objects.

**Comment by erikgrinaker** [comment]
@simen mentioned that `in path()` is basically a type conversion, such that `"a.b.c" in path("a.**")` typecasts `"a.b.c"` to a path and applies `in` with 2 path-parameters. Such implicit type casting is a feature of weakly typed systems, so changing to strong typing would require explicit typecasts, e.g. `path("a.b.c") in path("a.**")`, thus breaking backwards compatibility.

**Comment by simen** [comment]
I don't think we need `<` to be exactly the same `<` as used when sorting. As you mention, we can augment the `<` with a defined type order specifically for sorting. We have a defined order for types in `type_order.go`. Currently, it reads:

```
var TypeOrder = []datatypes.Type{datatypes.DateTime, datatypes.Float, datatypes.String, datatypes.Bool}
```

I don't think that is set in stone or even exhaustive. Don't know if it has predictable behavior for nulls.

**Comment by erikgrinaker** [comment]
Meeting summary: use strong typing, same types as JSON, interpret complex types (e.g. dates and refs) in code.
